plugins/ecc-vscc/ecc_validation_logic.go

```diff
@@ -121,70 +120,70 @@ func (vscc *VSCCECC) checkEnclaveEndorsement(cis *peer.ChaincodeInvocationSpec,
+	var txType []byte
+	var argss []string
+	if string(function) == "__init" {
+		txType = []byte("init")
+		argss = make([]string, len(cis.ChaincodeSpec.Input.Args[1:]))
+		for i, v := range cis.ChaincodeSpec.Input.Args[1:] { // drop "__init"
+			argss[i] = string(v)
 		}
-		var encoded_args []byte
-		encoded_args, err = json.Marshal(argss)
-		if err != nil {
-			return fmt.Errorf("Couldn't json encode arguments, err %s", err)
+	} else {
+		txType = []byte("invoke")
+		argss = make([]string, len(cis.ChaincodeSpec.Input.Args))
+		for i, v := range cis.ChaincodeSpec.Input.Args {
+			argss[i] = string(v)
 		}
-		logger.Debugf("txType: %s / encoded_args: %s\n", string(txType), string(encoded_args))
+	}
+	var encoded_args []byte
+	encoded_args, err = json.Marshal(argss)
+	if err != nil {
+		return fmt.Errorf("Couldn't json encode arguments, err %s", err)
+	}
+	logger.Debugf("txType: %s / encoded_args: %s\n", string(txType), string(encoded_args))
 
-		// get the enclave response
-		response := &sgx_utils.Response{}
-		if err := json.Unmarshal(respPayload.Response.Payload, response); err != nil {
-			return fmt.Errorf("Unmarshalling of SGX response failed, err: %s", err)
-		}
-		logger.Debugf("response: %s\n", string(response.ResponseData))
+	// get the enclave response
+	response := &utils.Response{}
+	if err := json.Unmarshal(respPayload.Response.Payload, response); err != nil {
+		return fmt.Errorf("Unmarshalling of SGX response failed, err: %s", err)
+	}
+	logger.Debugf("response: %s\n", string(response.ResponseData))
 
-		// check that response.PublicKey is registred
-		enclavePkHash := sha256.Sum256(response.PublicKey)
-		base64PublicKey := base64.StdEncoding.EncodeToString(enclavePkHash[:])
-		logger.Debugf("pk: %s", base64PublicKey)
+	// check that response.PublicKey is registred
+	enclavePkHash := sha256.Sum256(response.PublicKey)
+	base64PublicKey := base64.StdEncoding.EncodeToString(enclavePkHash[:])
+	logger.Debugf("pk: %s", base64PublicKey)
 
-		// FIXME: remove hardcoding of those strings
-		attestation, err := state.GetState("ercc", base64PublicKey)
-		if err != nil || attestation == nil {
-			return fmt.Errorf("Enclave PK not found in registry")
-		}
+	// FIXME: remove hardcoding of those strings
+	attestation, err := state.GetState("ercc", base64PublicKey)
+	if err != nil || attestation == nil {
+		return fmt.Errorf("Enclave PK not found in registry")
+	}
+	logger.Debugf("found attestation at ercc for pk: %s", base64PublicKey)
 
-		// Next, reproduce sorted read/writeset
-		var readset, writeset [][]byte
+	var readset, writeset [][]byte
+	// go through the read/writeset as produced within the enclave
+	for _, ns := range txRWSet.NsRwSets {
+		if ns.NameSpace != cis.ChaincodeSpec.ChaincodeId.Name {
+			// note that the read/write set may contain other namespaces than our fpc chaincode such as lifecycle
+			// we only care about FPC chaincode namespace and therefore filter
+			continue
+		}
```

Comment:
I would have expected that ercc would be the only other namespace?  Is a custom validation plugin also called on lifecycle events?  That would be kindof surprising (e.g., for chaincode spec updates updating plugins, which plugins are called?!) but more importantly, how would we make sure the corresponding transactions are properly verified!  By doing continue what would validate the corresponding path?   In fact, i doubt that lifecycle tx would come here as they usually would not have ecc attestations and hence below verification fails?

So ideally we would test that it is ercc or fail otherwise?  In any case, i think the current comment is wrong and should be updated

Reply:
here we iterate over the read/writeset of a transaction associated with ecc-vscc. As an ecc-enclave endorses only read/writes that are part of the fpc chaincode namespace, any other entry showing up here are not signed by the enclave, thus, must be filtered out in order to verify the enclave signature.

What are those other entries? If you look at the logs we can see that every chaincode transaction contains a readset entry to `_lifecycle`.`$chaincode_name$`. In particular, TLCC logs show it pretty well and I assume that explorer would show it as well.
It seems that this is a check that all endorsements use the same "chaincode version". Anyway, as this appears in the readset, the entry is covered by the peer endorsement which we have on top of ecc-vscc right now. So no need for additional checks here.

Reply:
oh, i see, never saw these __lifecyce values. So definitely was wrong on that.